setup.go

```diff
@@ -0,0 +1,826 @@
+// Copyright 2019 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+// +build !minimal
+
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/mendersoftware/log"
+	"github.com/mendersoftware/mender/client"
+
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+type setupOptionsType struct {
+	configPath         string
+	deviceType         string
+	username           string
+	password           string
+	serverURL          string
+	serverIP           string
+	serverCert         string
+	tenantToken        string
+	invPollInterval    int
+	retryPollInterval  int
+	updatePollInterval int
+	hostedMender       bool
+	demo               bool
+}
+
+// ------------------------------ Setup constants ------------------------------
+const ( // state enum
+	stateDeviceType = iota
+	stateDeviceTypeConfirm
+	stateHostedMender
+	stateDemoMode
+	stateServerURL
+	stateServerIP
+	stateServerCert
+	stateHostedMenderCredentials
+	statePolling
+	stateDone
+)
+
+const ( // Prompt constants
+	promptWizard = `Mender Client Setup
+===================
+
+Setting up the Mender client: the client runs on your device as a daemon, and will regularly poll the server to check for updates and report its inventory data.
+
+Get started by first configuring the device type and settings for communicating with the server:
+`
+	promptDone = "Starting Mender client…\n" + // NOTE: format
+		"Your device should show up at %s within the next few " +
+		"minutes.\nLog in to your Mender UI at %s to accept it.\n"
+
+	promptDeviceType   = "Device type (for example raspberrypi3-raspbian): "
+	promptConfirmDev   = "Confirm devicetype as %s? [Y/n] " // NOTE: format
+	promptHostedMender = "\nAre you connecting this device to Hosted Mender? [Y/n] "
+	promptCredentials  = "\nEnter your Hosted Mender credentials"
+	promptDemoMode     = "\nDemo mode uses short poll intervals and assumes the " +
+		"default demo server setup. (Recommended for testing.)\n" +
+		"Do you want to run the client in demo mode? [Y/n] "
+
+	promptServerIP   = "\nSet the IP of the Mender Server: [127.0.0.1] "
```

Comment:
On a somewhat unrelated note, if I may add my own opinion -- I would fuse the `stateServerIP` and `stateServerURL` to prompt for a valid email or IP address. Adding a local route to `/etc/hosts` enforces root privileges for running setup. By simply configuring the server IP in the `mender.conf:Servers:ServerURL` prefixed with the appropriate schema `http/https` we can achieve the same result without enforcing privileged access.

Anyway, +1 on adding `const` to all default values.

Reply:
When configuring a demo/non-hosted setup we ask for server IP, making a static route from the default url to the given IP in `/etc/hosts` and then setting the `Servers::ServerURL` in `mender.conf` to this default url (`https://docker.mender.io`). Instead, I would suggest that we treat this case the same as the non-demo case asking for the ServerURL but we allow the host field of the URL to take either a named server or an IP address (for instance https://127.0.0.1) and configuring this as the serverURL in `mender.conf`. In practice the two approaches produces the same result, but the latter doesn't necessarily require root privileges.

Reply:
Thanks for clarifying. Conceptually I like very much your idea, but I suspect that there might be some other reason for having this ugly hack in /etc/hosts that we are not considering here.

Can you check this with Kristian?

Reply:
@kacf; is there something I'm missing here? (see comments above)

Reply:
Yes, IP doesn't work, because TLS handshake requires a host name, and the host name has to be the same as the one you generated the certificate for. The demo certificate we ship is signed for `docker.mender.io`.

Reply:
Ohh, right. Then I guess we'll leave it as it is.